Issue 253 repeatable shifts

Taktar opened the pull request:
Pull request for Issue 253 repeatable shifts 

What was changed?
Added a new route for defining repeatable shifts (/shelter-dashboard/:shelterId/repeatable-shifts).

Updated the Settings.js file to include a button link that navigates to the Repeatable Shifts page.

Created a placeholder for the RepeatableShifts component (if you did).

Why was it changed?
Shelter admins need the ability to define repeatable shifts easily from the dashboard.

This functionality helps streamline shift scheduling and reduces manual entry for common daily schedules.

How was it changed?
Modified client_app/src/components/shelter/Settings.js:

Added a <Link> that routes to /shelter-dashboard/:shelterId/repeatable-shifts with a Define Repeatable Shifts button.

Modified client_app/src/App.js:

Imported RepeatableShifts and added a new <Route> inside the <ShelterDashboardLayout> route group.

Taktar commented:
<img width="1439" alt="Screenshot 2025-04-28 at 3 25 11 PM" src="https://github.com/user-attachments/assets/5872521e-02a6-4713-bab6-fda18fc81866" />
<img width="966" alt="Screenshot 2025-04-28 at 3 25 33 PM" src="https://github.com/user-attachments/assets/f67f42d3-a427-4884-9ece-724b63a7fa50" />
<img width="1016" alt="Screenshot 2025-04-28 at 3 25 49 PM" src="https://github.com/user-attachments/assets/e23a6ba1-5ba8-4297-bd17-850d3fa039a5" />
when clicking the clock icon, you can set up the start and end time for a shift.

Taktar commented:
<img width="865" alt="Screenshot 2025-04-28 at 3 28 00 PM" src="https://github.com/user-attachments/assets/2d163b97-d3bd-4a04-b62c-89d92fced9dd" />
Once I specify the day, like here I typed in Monday, it allows me to add the shift successfully! I was also able to delete the Monday shift after creating it!

Taktar commented on client_app/src/components/shelter/RepeatableShifts.js:
You have to type in a day for it to work

Taktar replied:
Try typing "Monday" in the area above the start time

Taktar commented:
Hi! For some reason I am unable to upload a picture or screenshot of my work into github, so I wanted to send this email with an attachment for now!

I added features based on your feedback.

[Saint Louis University]

Taslima Aktar she/her
Saint Louis University Class of 2025
College of Arts and Sciences
Computer Science Major


Phone: 314-398-5048

***@***.***
slu.edu<https://www.slu.edu/> | slubillikens.com<https://slubillikens.com/>

________________________________
From: Kate Holdener ***@***.***>
Sent: Monday, April 28, 2025 5:35 PM
To: oss-slu/shelter_volunteers ***@***.***>
Cc: Taslima Aktar ***@***.***>; Assign ***@***.***>
Subject: [External] Re: [oss-slu/shelter_volunteers] Issue 253 repeatable shifts (PR #265)

@ kate-holdener requested changes on this pull request. In client_app/src/components/shelter/RepeatableShifts. js: > @@ -0,0 +1,110 @@ +// client_app/src/components/shelter/RepeatableShifts. js + +import React, { useState } from "react";


@kate-holdener requested changes on this pull request.

________________________________

In client_app/src/components/shelter/RepeatableShifts.js<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/265*discussion_r2064920832__;Iw!!K543PA!P1Cqdwf_ny-bW4vl0jklDuQdo-xFdYJ8EkKYglWdZL2KjF_-6hAECZlOmpocZWS1dNN8sVXoy5hVQAJZrSWm4ghlt1M$>:

> @@ -0,0 +1,110 @@
+// client_app/src/components/shelter/RepeatableShifts.js
+
+import React, { useState } from "react";
+import { useParams } from "react-router-dom";
+
+const RepeatableShifts = () => {
+  const { shelterId } = useParams(); // grab the shelterId from URL
+
+  const [shifts, setShifts] = useState([]);
+  const [newShift, setNewShift] = useState({
+    day: "",


We don't want to specify the day in these repeatable shifts.
We will want to allow the user to specify:

  *   shift_start (store that as milliseconds since midnight)
  *   shift_end (store that as milliseconds since midnight)
  *   shift name (optional)
  *   required volunteer count (optional, defaulted to 1)
  *   max volunteer count (optional, defaulted to 5)

________________________________

On client_app/src/components/shelter/RepeatableShifts.js<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/265*discussion_r2064929979__;Iw!!K543PA!P1Cqdwf_ny-bW4vl0jklDuQdo-xFdYJ8EkKYglWdZL2KjF_-6hAECZlOmpocZWS1dNN8sVXoy5hVQAJZrSWmjmbZyg0$>:

When I click on "Add Shifts" button, nothing happens. Also, we want to allow the user to add a bunch of shifts and then submit them. There is no way to submit shifts.
Screenshot.2025-04-28.at.5.33.08.PM.png (view on web)<https://urldefense.com/v3/__https://github.com/user-attachments/assets/60f1c0f9-4fe9-4579-b764-6b0e8114544f__;!!K543PA!P1Cqdwf_ny-bW4vl0jklDuQdo-xFdYJ8EkKYglWdZL2KjF_-6hAECZlOmpocZWS1dNN8sVXoy5hVQAJZrSWmtkcTRig$>

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/265*pullrequestreview-2801175560__;Iw!!K543PA!P1Cqdwf_ny-bW4vl0jklDuQdo-xFdYJ8EkKYglWdZL2KjF_-6hAECZlOmpocZWS1dNN8sVXoy5hVQAJZrSWmCkdt68Q$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFOPC4BB2SY5EYYUUSUOIPT232UKHAVCNFSM6AAAAAB4BGEMUCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQMBRGE3TKNJWGA__;!!K543PA!P1Cqdwf_ny-bW4vl0jklDuQdo-xFdYJ8EkKYglWdZL2KjF_-6hAECZlOmpocZWS1dNN8sVXoy5hVQAJZrSWm6IQyQ-s$>.
You are receiving this because you were assigned.Message ID: ***@***.***>

Taktar commented:
<img width="1427" alt="Screenshot 2025-04-28 at 11 16 02 PM" src="https://github.com/user-attachments/assets/991824ee-abfc-4ef4-99cb-28167a5d1829" />
cleaner look and I can add as many shifts as I want and delete them as well.

kate-holdener commented on client_app/src/components/shelter/RepeatableShifts.js:
```diff
@@ -0,0 +1,195 @@
+        requiredVolunteers: 1,
+        maxVolunteers: 5,
+      });
+      setSuccessMessage("Shift added successfully!");
+      setTimeout(() => setSuccessMessage(""), 3000); // Remove after 3 sec
```
I think the success message here is a bit early, since the shifts are just added to the list to be submitted (they are not submitted to the backend yet).

kate-holdener commented on client_app/src/components/shelter/RepeatableShifts.js:
```diff
@@ -0,0 +1,195 @@
+  };
+
+  return (
+    <div className="repeatable-shifts-page" style={{ maxWidth: "900px", margin: "0 auto", padding: "20px" }}>
+      <h2>Define Repeatable Shifts for Shelter {shelterId}</h2>
+      <form onSubmit={handleSubmit} style={{ marginBottom: "20px" }}>
```
Let's move all the styling elements into the .css file

kate-holdener commented on client_app/src/components/shelter/RepeatableShifts.js:
```diff
@@ -0,0 +1,195 @@
+          />
+        </label>
+        <br />
+        <button type="submit" style={{ marginTop: "10px" }}>
+          Add Shift
+        </button>
+        {successMessage && <p style={{ color: "green", marginTop: "10px" }}>{successMessage}</p>}
```
More styling elements that should be moved to css

kate-holdener reviewed (changes requested):
This is great progress, @Taktar! I added a couple of comments.
The user can build up a list of repeatable shifts. They still need a way to submit those shifts for storage on the backend. You can add "Submit" a button at the bottom of the list. Also, it would be good to include brief instructions at the top, telling the user that this page is used to build a schedule that will be used when they open a shelter. Tell the user to create the shifts they need and then submit them. 

You will need to work with @amalrzv and @emrameduseljac to integrate this with the backend API.

On the front end, we have src/api directory, where we have a wrapper for all backend API calls. You will want to add a schedule.js file there, and define a function that you would call to post the schedule. I believe @strejoarciles added such a file in her branch. You may want to work with her to the front end changes into the same branch: either you can merge your canges to @strejoarciles branch or vice versa.

Taktar commented:
All feedback addressed: moved inline styles to CSS, delayed success message until submission, added instructional text, submit button added.
<img width="1436" alt="Screenshot 2025-05-04 at 1 14 06 PM" src="https://github.com/user-attachments/assets/ecd2e8a9-1213-4138-9209-209df5209fef" />

"Schedule submitted successfully" shows when the submit button is clicked.
